tests/dialects/test_hw.py

```diff
@@ -25,3 +46,132 @@ def test_inner_sym_target():
+class ModuleOp(IRDLOperation):
+    name = "module"
+    region: Region = region_def()
+    sym_name = attr_def(StringAttr)
+    traits = frozenset({InnerSymbolTable(), SymbolOpInterface()})
+
+
+@irdl_op_definition
+class OutputOp(IRDLOperation):
+    name = "output"
+    traits = frozenset({IsTerminator()})
+
+
+@irdl_op_definition
+class CircuitOp(IRDLOperation):
+    name = "circuit"
+    region: Region | None = opt_region_def()
+    sym_name = attr_def(StringAttr)
+    traits = frozenset(
+        {InnerRefNamespace(), SymbolTable(), SingleBlockImplicitTerminator(OutputOp)}
+    )
+
+    def __post_init__(self):
+        for trait in self.get_traits_of_type(SingleBlockImplicitTerminator):
+            ensure_terminator(self, trait)
+
+
+def test_inner_symbol_table_interface():
+    """
+    Test that operations that conform to SymbolOpInterface have necessary attributes.
+    """
+
+    @irdl_op_definition
+    class MissingTraitModuleOp(IRDLOperation):
+        name = "module"
+        traits = frozenset({InnerSymbolTable()})
+
+    op0 = MissingTraitModuleOp()
+    iface0 = op0.get_trait(InnerSymbolTable)
+
+    assert iface0 is not None
+    with pytest.raises(
+        VerifyException, match="Operation module must have trait SymbolOpInterface"
+    ):
+        # NB. seems to be called in a separate pass
+        iface0.verify_region_trait(op0)
+
+    @irdl_op_definition
+    class MissingAttrModuleOp(IRDLOperation):
+        name = "module"
+        traits = frozenset({InnerSymbolTable(), SymbolOpInterface()})
+
+    op1 = MissingAttrModuleOp()
+
+    with pytest.raises(
+        VerifyException,
+        match='Operation module must have a "sym_name" attribute of type `StringAttr`',
+    ):
+        op1.verify()
+
+    # Using the globally-defined ModuleOp
+    op2 = ModuleOp(
+        attributes={"sym_name": StringAttr("symbol_name")}, regions=[[OutputOp()]]
+    )
+    op2.verify()
+    iface2 = op2.get_trait(InnerSymbolTable)
+
+    @irdl_op_definition
+    class ParentOp(IRDLOperation):
+        name = "parent"
+        region: Region = region_def()
+
+    ParentOp(regions=[[op2]])
+
+    assert iface2 is not None
+    with pytest.raises(
+        VerifyException,
+        match="Operation module with trait InnerSymbolTable must have a parent with trait InnerRefNamespace",
+    ):
+        iface2.verify_region_trait(op2)
+
+    # Using the globally-defined ModuleOp again
+    op4 = ModuleOp(
+        attributes={"sym_name": StringAttr("symbol_name")}, regions=[[OutputOp()]]
+    )
+    op5 = CircuitOp(attributes={"sym_name": StringAttr("other_name")}, regions=[[op4]])
+
+    op5.verify()
+    iface3 = op4.get_trait(InnerSymbolTable)
+    assert iface3 is not None
+    iface3.verify_region_trait(op4)
+
+
+@irdl_op_definition
+class InnerRefUserOp(IRDLOperation):
+    name = "ref"
+    sym_name = attr_def(StringAttr)
+    traits = frozenset({InnerRefUserOpInterface()})
+
+
+@irdl_op_definition
+class WireOp(IRDLOperation):
```

Comment:
Maybe move this up in the file to be together with the rest of the "mock" classes?